A7-A11, T2 SoC cpufreq support

asdfugil opened the pull request:
This adds the tunables required for cpufreq to work properly on A7-A11, T2 SoCs, most prominently a register poke to enable voltage changes when the state is changed, as iBoot leaves the voltage controls disabled.

asdfugil committed:
pmgr: Add pmgr_power_on() function to power on power domains by name

Some power domains are not referenced by other nodes through clock-gates
or power-gates, so this is needed.

Signed-off-by: Nick Chan <towinchenmi@gmail.com>

asdfugil committed:
utils: Add poll64() function

This is the 64-bit variant of poll32().

Signed-off-by: Nick Chan <towinchenmi@gmail.com>

asdfugil commented:
this is also a requirement for proper operation of the proxy with the physical uart, as the cpu is too slow otherwise

svenpeter42 commented on src/cpufreq.c:
```diff
@@ -39,23 +44,41 @@ struct feat_t {
-        val |= CLUSTER_PSTATE_SET | FIELD_PREP(CLUSTER_PSTATE_DESIRED1, pstate);
-        if (chip_id == T8103 || chip_id <= T6002) {
+    if (get_target_pstate(cluster) != pstate) {
+        if (chip_id == S5L8960X || chip_id == T7000 || chip_id == T7001) {
```
I'd just use a switch/case with all chip_id listed explicitly here even if that results in some code duplication. Even the chip_id <= T6002 in the original code is a bit weird but now that we have three(?) different variants this is getting confusing.

svenpeter42 commented on src/cpufreq.c:
```diff
@@ -89,11 +112,46 @@ int cpufreq_init_cluster(const struct cluster_t *cluster, const struct feat_t *f
 
     /* Unknown */
-    write64(cluster->base + 0x440f8, 1);
+    if ((chip_id > T8015 && chip_id < S5L8960X) || chip_id < T7000)
```
I'd really avoid all those > and < operations on chip_id and just hardcode a list where we need to do this write instead like e.g. the switch below

asdfugil committed:
cpufreq: refractor code to use hardcoded chip_id operations

The existing code is already confusing as is and support for more chips with
different requirements will be added, which would make the code even more
confusing. Instead, switch to a hardcoded list of chip_id's for operations
that are chip-specific or will be chip-speciifc.

Signed-off-by: Nick Chan <towinchenmi@gmail.com>

asdfugil committed:
cpufreq: Switch to poll64() for polling CLUSTER_PSTATE

cpufreq support for Apple A9 will be added. It requires 64-bit access on the
CLUSTER_PSTATE register.

Signed-off-by: Nick Chan <towinchenmi@gmail.com>

asdfugil committed:
cpufreq: Add Apple A7-A11, T2 SoC support

- A7-A8X uses 3 bits only for pstate, and PS1 mask is GENMASK(24, 22)

- Add tunables, mainly the voltage control register in PMGR (iBoot boots us
with voltage changes seemingly disabled, with obvious disastrous outcome
trying to switch to high P-states)

- Skip writes to offset 0x4xxxx from cluster->base for A7-A11, T2. The range
starting from 0x40000 from cluster->base does not exist in ADT /arm-io/pmgr,
so it probably does not exist for those older chips. This includes the
0x440f8 and ppt-thrtl writes.

Signed-off-by: Nick Chan <towinchenmi@gmail.com>

asdfugil commented:
Changes:
- Switch to `switch (chip_id)` for chip-specific operations, this removes all the comparison operations.

svenpeter42 commented on src/cpufreq.c:
```diff
@@ -39,19 +44,44 @@ struct feat_t {
+    if (pstate_reg_to_pstate(val) != pstate) {
+        switch (chip_id) {
+            case S5L8960X:
+            case T7000 ... T7001:
```
This is now just hiding the comparisons that I'd like to avoid. Please get rid of the ... stuff as well and just open code all the chip_ids here.

svenpeter42 commented on src/cpufreq.c:
```diff
@@ -39,19 +44,44 @@ struct feat_t {
+            case S8000 ... T8103:
+            case T6000 ... T6002:
+                val &= ~CLUSTER_PSTATE_DESIRED2;
+                val |= CLUSTER_PSTATE_SET | FIELD_PREP(CLUSTER_PSTATE_DESIRED2, pstate);
```
please add a // fallthrough comment here

svenpeter42 commented on src/cpufreq.c:
```diff
@@ -88,11 +118,44 @@ int cpufreq_init_cluster(const struct cluster_t *cluster, const struct feat_t *f
+        case S5L8960X:
+            write32(pmgr_reg + PMGR_VOLTAGE_CTL_OFF_S5L8960X, 1);
+            break;
+        case T7000 ... T7001:
```
same as above: please don't rely on hidden comparisons here and just enumerate all chips

svenpeter42 commented on src/cpufreq.c:
```diff
@@ -88,11 +118,44 @@ int cpufreq_init_cluster(const struct cluster_t *cluster, const struct feat_t *f
+            break;
+        case T8103 ... T8122:
+        case T6000 ... T6034:
+        default:
```
Please also separate out the default case and just print that the chip is unsupported

asdfugil commented:
Changes:
- Remove *all* hidden comparison using `switch (chip_id)`, including the existing ones, now every supported chip is explicitly stated.
- Don't "support" T8122 and T6034
- Print chip is unsupported in default cases
